core/tests/vm-benchmark/benches/instructions.rs

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+}
+
+impl Reporter for MetricsReporter {
+    fn new_benchmark(&mut self, id: &BenchmarkId) -> Box<dyn BenchmarkReporter> {
+        Box::new(MetricsBenchmarkReporter(id.clone()))
+    }
+}
+
+#[derive(Debug)]
+struct MetricsBenchmarkReporter(BenchmarkId);
+
+impl BenchmarkReporter for MetricsBenchmarkReporter {
+    fn ok(self: Box<Self>, output: &BenchmarkOutput) {
+        #[derive(Debug, Metrics)]
+        #[metrics(prefix = "vm_cachegrind")]
+        struct VmCachegrindMetrics {
+            #[metrics(labels = ["benchmark"])]
+            instructions: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l1_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l2_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            ram_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            cycles: LabeledFamily<String, Gauge<u64>>,
+        }
+
+        #[vise::register]
+        static VM_CACHEGRIND_METRICS: vise::Global<VmCachegrindMetrics> = vise::Global::new();
+
+        let id = self.0.to_string();
+        VM_CACHEGRIND_METRICS.instructions[&id].set(output.stats.total_instructions());
+        if let Some(&full) = output.stats.as_full() {
+            let summary = AccessSummary::from(full);
+            VM_CACHEGRIND_METRICS.l1_accesses[&id].set(summary.l1_hits);
+            VM_CACHEGRIND_METRICS.l2_accesses[&id].set(summary.l3_hits);
+            VM_CACHEGRIND_METRICS.ram_accesses[&id].set(summary.ram_accesses);
+            VM_CACHEGRIND_METRICS.cycles[&id].set(summary.estimated_cycles());
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Comparison {
+    current_cycles: u64,
+    prev_cycles: Option<u64>,
+}
+
+impl Comparison {
+    fn percent_difference(a: u64, b: u64) -> f64 {
+        ((b as f64) - (a as f64)) / (a as f64) * 100.0
+    }
+
+    fn new(output: &BenchmarkOutput) -> Option<Self> {
+        let current_cycles = AccessSummary::from(*output.stats.as_full()?).estimated_cycles();
+        let prev_cycles = if let Some(prev_stats) = &output.prev_stats {
+            Some(AccessSummary::from(*prev_stats.as_full()?).estimated_cycles())
+        } else {
+            None
+        };
+
+        Some(Self {
+            current_cycles,
+            prev_cycles,
+        })
+    }
+
+    fn cycles_diff(&self) -> Option<f64> {
+        self.prev_cycles
+            .map(|prev_cycles| Self::percent_difference(prev_cycles, self.current_cycles))
+    }
+}
+
+/// Reporter that outputs diffs in a Markdown table to stdout after all benchmarks are completed.
+///
+/// Significant diff level can be changed via `BENCHMARK_DIFF_THRESHOLD_PERCENT` env var; it is set to 1% by default.
+#[derive(Debug, Default)]
+struct ComparisonReporter {
+    comparisons: Arc<Mutex<BTreeMap<String, Comparison>>>,
```

Comment:
I'd use a queue over a mutex here.
https://doc.rust-lang.org/std/sync/mpsc/index.html

Reply:
No objections, but just to clarify: are you concerned about contention? IMO, it's not much of an issue in this particular case since the number of test cases is small.

Reply:
No, I just prefer message passing over low-level constructs.